examples: provide example application for a 6LoWPAN border router

authmillenon opened the pull request:
Needs ~~#3750~~ (merged) for context configuration.

OlegHahm commented on examples/gnrc_border_router/Makefile:
```diff
@@ -0,0 +1,48 @@
+  SLIP_BAUDRATE=115200
+endif
+
+CFLAGS += -I$(CURDIR)
```
INCLUDES?

OlegHahm commented on examples/gnrc_border_router/Makefile:
```diff
@@ -0,0 +1,48 @@
+# name of your application
+APPLICATION = gnrc_networking
```
gnrc_border_router

OlegHahm commented on examples/gnrc_border_router/Makefile:
```diff
@@ -0,0 +1,48 @@
+# Specify the mandatory networking modules for 6LoWPAN border router
+USEMODULE += gnrc_sixlowpan_border_router_default
+# Add a routing protocol
+USEMODULE += gnrc_rpl
```
To keep it slim, maybe we should omit this. (Maybe replace by fib.)

authmillenon replied:
FIB is pulled in automatically and I thought the border router is supposed to be root

OlegHahm replied:
Yes, in RPL network the 6LBR would usually act as a root, but for a simple 6LBR example we could just skip RPL.

authmillenon replied:
I was wrong FIB is not pulled in by the border router. Included FIB now.

OlegHahm commented:
I guess a small README would be helpful.

authmillenon commented:
> I guess a small README would be helpful.

I thought it might be better if you provide it, since you already worked with it and aren't the person who wrote it.

gebart commented:
Does this border router have any external scripts for setting up the SLIP connection towards the PC? Does tunslip6 which we brought in a few months ago work out of the box with this or does it require any extra RIOT shell init commands to be executed first to bring the network up?

OlegHahm commented:
@gebart, ok, I will take care of these questions in the README. But I will try to fix the 6LBR first, before I write the README. ;)

authmillenon commented on examples/gnrc_border_router/Makefile:
```diff
@@ -0,0 +1,48 @@
+APPLICATION = gnrc_border_router
+
+# If no BOARD is found in the environment, use this default:
+BOARD ?= samr21-xpro
```
I changed this to samr21-xpro since it is the board we tested on and since native has no support for `periph/uart`.

OlegHahm replied:
Makes sense.

OlegHahm commented:
Please squash.

authmillenon commented:
Squashed

kb2ma commented on examples/gnrc_border_router/Makefile:
```diff
@@ -0,0 +1,48 @@
+  # set default
+  SLIP_BAUDRATE=115200
+endif
+
```
Should be 'ifeq' in l. 14, right?

kb2ma commented:
Makefile, l. 14 should be 'ifeq', not 'ifneq' for BAUDRATE, right? Sorry, I don't know how to quote in the PR. :-/

authmillenon commented:
@kb2ma 

> Makefile, l. 14 should be 'ifeq', not 'ifneq' for BAUDRATE, right?

You're right, thanks! Changed and squashed immediately.

> Sorry, I don't know how to quote in the PR. :-/

Just hit the blue plus, when you mouse-over a line in the diff.

gebart commented:
@kb2ma and anyone else who is not used to Github

>Just hit the blue plus, when you mouse-over a line in the diff.

To clarify, click on the "Files changed (3)" tab at the top of this page, and write your comments on that page. When you do that the comment along with a few lines of code preceding the comment will be seen from inside the Comments tab (this page) as well.

(The alternative, and not recommended, way is to go to the commits tab and click on a commit and write the comments on that page. However, those comments will not be visible from inside this Comment page, which makes it very easy to miss. Please try to avoid using this alternative method when discussing a pull request)

OlegHahm commented on examples/gnrc_border_router/Makefile:
```diff
@@ -0,0 +1,48 @@
+# Specify the mandatory networking modules for 6LoWPAN border router
+USEMODULE += gnrc_sixlowpan_border_router_default
+# Add a routing protocol
+USEMODULE += gnrc_fib
```
`s/gnrc_fib/fib`

authmillenon replied:
Done

kb2ma commented:
@authmillenon @gebart 
Thanks for the tip and the distinction between 'Files changed' and 'commits'. Very clear.

OlegHahm commented:
First draft of a README in https://github.com/authmillenon/RIOT/pull/24

kb2ma commented:
When I add the line below to Makefile

CFLAGS += -DGNRC_NETIF_NUMOF=2

I can ping now via slip between samr21_xpro LBR and Linux/TUN! I was wondering where interface 6 came from in Oleg's README. :-)

OlegHahm commented:
Hehe, nice coincident, I was just about to comment that exactly this line is missing in the Makefile. ;)

authmillenon commented:
Fixed

authmillenon commented:
Merged @OlegHahm's README

gebart commented on examples/gnrc_border_router/README.md:
```diff
@@ -0,0 +1,62 @@
+offers an IPv6 capable network interface (e.g. the `encx24j600` Ethernet chip)
+or connect it over the serial interface to a Linux host and use the SLIP
+standard [1]. The example application in this folder assumes as a default to be
+run on an Atmel SAM R21 Xplained Pro evaluation using an external UART adapter
```
..."evaluation **board**"...

gebart commented on examples/gnrc_border_router/main.c:
```diff
@@ -0,0 +1,36 @@
+
+int main(void)
+{
+    puts("RIOT network stack example application");
```
Change to say "border router" or something?

authmillenon commented:
Addressed comments

kb2ma commented:
Should the Makefile include the gnrc_rpl module for routing rather than fib?

authmillenon commented:
> Should the Makefile include the gnrc_rpl module for routing rather than fib?

https://github.com/RIOT-OS/RIOT/pull/3886#discussion_r39739377

kb2ma commented:
Thanks; sorry for the oversight. I need to step back and dig a little deeper.

authmillenon commented:
No problem. :-)

OlegHahm commented:
Please squash!

authmillenon commented:
Squashed

authmillenon commented:
I was trying to retrigger travis, but it seems to be sleeping...

authmillenon commented:
Rebased to current master.

authmillenon commented:
Adapted for boards with insuffient RAM and squashed immediately. For the AVR error there will come a seperate PR.

authmillenon commented:
See #3904

authmillenon commented:
Forget about the dependency. `arduino-mega2560` doesn't have `core_panic()` defined, making compiling this not possible.
